crates/task-impls/src/request.rs

```diff
@@ -219,140 +218,117 @@ impl<TYPES: NodeType, I: NodeImplementation<TYPES>> NetworkRequestState<TYPES, I
-        }
+        });
+        self.spawned_tasks.entry(view).or_default().push(handle);
     }
-    /// Handle sending a VID Share request, runs the loop until the data exists
-    async fn do_vid(&self, req: VidRequest<TYPES>, signature: Signature<TYPES>) {
-        let message = make_vid(&req, signature);
-        let mut recipients_it = self.recipients.iter().cycle();
-
-        let serialized_msg = match bincode::serialize(&message) {
-            Ok(serialized_msg) => serialized_msg,
-            Err(e) => {
-                tracing::error!(
-                    "Failed to serialize outgoing message: this should never happen. Error: {e}"
-                );
-
-                return;
-            }
-        };
 
-        while !self.cancel_vid(&req).await {
-            match async_timeout(
-                REQUEST_TIMEOUT,
-                self.network
-                    .request_data::<TYPES>(serialized_msg.clone(), recipients_it.next().unwrap()),
-            )
-            .await
-            {
-                Ok(Ok(response)) => {
-                    match bincode::deserialize(&response) {
-                        Ok(ResponseMessage::Found(data)) => {
-                            self.handle_response_message(data).await;
-                            // keep trying, but expect the map to be populated, or view to increase
-                            async_sleep(REQUEST_TIMEOUT).await;
-                        }
-                        Ok(ResponseMessage::NotFound) => {
-                            info!("Peer Responded they did not have the data");
-                        }
-                        Ok(ResponseMessage::Denied) => {
-                            error!("Request for data was denied by the receiver");
+    /// Wait for the response after we send out the request
+    /// Returns true if we get a response, otherwise false
+    async fn handle_response(
+        receiver: &Receiver<Arc<HotShotEvent<TYPES>>>,
+        sender: &Sender<Arc<HotShotEvent<TYPES>>>,
+        da_membership: &TYPES::Membership,
+        view: TYPES::Time,
+    ) -> bool {
+        let result = async_timeout(REQUEST_TIMEOUT, async {
+            let mut response = None;
+            while response.is_none() {
+                let event = EventDependency::new(
+                    receiver.clone(),
+                    Box::new(move |event: &Arc<HotShotEvent<TYPES>>| {
+                        let event = event.as_ref();
+                        if let HotShotEvent::VidResponseRecv(_sender_key, proposal) = event {
+                            proposal.data.view_number() == view
+                        } else {
+                            false
                         }
-                        Err(e) => {
-                            error!("Failed to deserialize response: {e}");
+                    }),
+                )
+                .completed()
+                .await;
+
+                if let Some(hs_event) = event.as_ref() {
```

Comment:
Why can't this check be part of the callback in the event dependency above.  If it were you could just create the dependency and await it and get rid of the while loop.